File: ios-native/Dimo/Auth/WorkOSAuthProvider.swift
import AuthenticationServices
import Combine
import ConvexMobile
import Foundation
import UIKit

final class WorkOSAuthProvider: NSObject, AuthProvider, @unchecked Sendable {
  typealias T = WorkOSSession

  private let refreshAccount = "workos.refreshToken"
  private let userAccount = "workos.user"
  private let lastProviderKey = "dimo.auth.lastProvider"
  private let lock = NSLock()
  private var cached: WorkOSSession?
  private var onIdToken: (@Sendable (String?) -> Void)?
  /// Single-flight refresh so concurrent TokenRefresher + cold-start paths
  /// cannot race WorkOS refresh-token rotation past the 30s grace window.
  private var refreshTask: Task<WorkOSSession, Error>?

  /// Provider used for the most recent successful sign-in, so a fresh `login`
  /// does not silently force Google on someone who signed up with Apple.
  private var lastProvider: String {
    get { UserDefaults.standard.string(forKey: lastProviderKey) ?? "GoogleOAuth" }
    set { UserDefaults.standard.set(newValue, forKey: lastProviderKey) }
  }

  var currentAccessToken: String? {
    lock.withLock { cached?.accessToken }
  }

  var currentSession: WorkOSSession? {
    lock.withLock { cached }
  }

  var hasPersistedRefreshToken: Bool {
    KeychainStore.get(account: refreshAccount) != nil
  }

  /// Cold-start restore. Retries transient network failures and only clears
  /// Keychain on a terminal `invalid_grant` (session truly over).
  func restoreSession() async -> WorkOSSession? {
    guard hasPersistedRefreshToken else { return nil }

    var lastTransient: Error?
    for attempt in 0..<5 {
      do {
        return try await loginFromCache(onIdToken: { _ in })
      } catch let error as AuthError where error.isTerminalRefresh {
        // Session revoked / expired at WorkOS — must re-authenticate.
        clearPersisted()
        return nil
      } catch {
        if AuthError.isTransient(error) || error is URLError {
          lastTransient = error
          let delay = min(8.0, pow(2.0, Double(attempt)))
          try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
          continue
        }
        // Unknown errors: keep the refresh token so a later launch can recover.
        lastTransient = error
        break
      }
    }

    // Still have a refresh token but WorkOS / network was unreachable. Keep
    // Keychain so the next launch can succeed — do not force a false logout.
    _ = lastTransient
    return nil
  }

  func signIn(provider: String) async throws -> WorkOSSession {
    try await performSignIn(provider: provider)
  }

  func signOut() async {
    try? await logout()
  }

  func refreshIfNeeded(force: Bool = false) async throws -> WorkOSSession {
    if !force, let current = lock.withLock({ cached }),
       current.expiresAt.timeIntervalSinceNow > 60 {
      return current
    }
    return try await refreshSingleFlight(force: force)
  }

  // MARK: AuthProvider

  func login(onIdToken: @Sendable @escaping (String?) -> Void) async throws -> WorkOSSession {
    self.onIdToken = onIdToken
    let session = try await performSignIn(provider: lastProvider)
    onIdToken(session.accessToken)
    return session
  }

  func logout() async throws {
    clearPersisted()
    onIdToken?(nil)
    onIdToken = nil
  }

  func loginFromCache(onIdToken: @Sendable @escaping (String?) -> Void) async throws -> WorkOSSession {
    self.onIdToken = onIdToken
    // Reuse a still-valid session from restore/sign-in so Convex auth does not
    // block on a second WorkOS refresh during every cold start.
    if let current = lock.withLock({ cached }),
       current.expiresAt.timeIntervalSinceNow > 60 {
      onIdToken(current.accessToken)
      return current
    }
    let session = try await refreshSingleFlight(force: true)
    onIdToken(session.accessToken)
    return session
  }

  func extractIdToken(from authResult: WorkOSSession) -> String {
    authResult.accessToken
  }

  // MARK: Private

  private func refreshSingleFlight(force: Bool) async throws -> WorkOSSession {
    if !force, let current = lock.withLock({ cached }),
       current.expiresAt.timeIntervalSinceNow > 60 {
      return current
    }

    let existing = lock.withLock { refreshTask }
    if let existing {
      return try await existing.value
    }

    let task = Task<WorkOSSession, Error> { [weak self] in
      guard let self else { throw AuthError.notAuthenticated }
      defer { self.lock.withLock { self.refreshTask = nil } }

      let refreshToken: String
      if let cachedToken = self.lock.withLock({ self.cached?.refreshToken }) {
        refreshToken = cachedToken
      } else if let stored = KeychainStore.get(account: self.refreshAccount) {
        refreshToken = stored
      } else {
        throw AuthError.notAuthenticated
      }

      do {
        let session = try await WorkOSAPI.refresh(
          refreshToken: refreshToken,
          clientId: AppConfig.workOSClientID
        )
        try self.persist(session)
        self.onIdToken?(session.accessToken)
        return session
      } catch let error as AuthError where error.isTerminalRefresh {
        self.clearPersisted()
        throw error
      }
    }

    lock.withLock { refreshTask = task }
    return try await task.value
  }

  private func performSignIn(provider: String) async throws -> WorkOSSession {
    let verifier = PKCE.makeVerifier()
    let challenge = PKCE.challenge(for: verifier)
    let state = PKCE.makeState()
    let url = WorkOSAPI.authorizationURL(
      clientId: AppConfig.workOSClientID,
      redirectURI: AppConfig.workOSRedirectURI,
      state: state,
      codeChallenge: challenge,
      provider: provider
    )
    let callback = try await startWebAuth(url: url)
    guard let components = URLComponents(url: callback, resolvingAgainstBaseURL: false) else {
      throw AuthError.missingCode
    }
    let items = Dictionary(uniqueKeysWithValues: (components.queryItems ?? []).compactMap { item in
      item.value.map { (item.name, $0) }
    })
    guard items["state"] == state else { throw AuthError.stateMismatch }
    guard let code = items["code"] else { throw AuthError.missingCode }
    let session = try await WorkOSAPI.exchangeCode(
      code: code,
      codeVerifier: verifier,
      clientId: AppConfig.workOSClientID,
      redirectURI: AppConfig.workOSRedirectURI
    )
    try persist(session)
    lastProvider = provider
    return session
  }

  private func persist(_ session: WorkOSSession) throws {
    // Persist the rotated refresh token before publishing the session so a
    // crash mid-refresh cannot leave the retired token as the only copy.
    try KeychainStore.set(session.refreshToken, account: refreshAccount)
    let userData = try JSONEncoder().encode(session.user)
    try KeychainStore.set(String(data: userData, encoding: .utf8) ?? "", account: userAccount)
    lock.withLock { cached = session }
  }

  private func clearPersisted() {
    KeychainStore.delete(account: refreshAccount)
    KeychainStore.delete(account: userAccount)
    UserDefaults.standard.removeObject(forKey: lastProviderKey)
    lock.withLock {
      cached = nil
      refreshTask?.cancel()
      refreshTask = nil
    }
  }

  @MainActor
  private func startWebAuth(url: URL) async throws -> URL {
    try await withCheckedThrowingContinuation { continuation in
      let session = ASWebAuthenticationSession(
        url: url,
        callbackURLScheme: "dimo"
      ) { callbackURL, error in
        if let error {
          let ns = error as NSError
          if ns.domain == ASWebAuthenticationSessionErrorDomain,
             ns.code == ASWebAuthenticationSessionError.canceledLogin.rawValue {
            continuation.resume(throwing: AuthError.cancelled)
          } else {
            continuation.resume(throwing: error)
          }
          return
        }
        guard let callbackURL else {
          continuation.resume(throwing: AuthError.missingCode)
          return
        }
        continuation.resume(returning: callbackURL)
      }
      session.presentationContextProvider = self
      session.prefersEphemeralWebBrowserSession = false
      if !session.start() {
        continuation.resume(throwing: AuthError.server("Could not start authentication session"))
      }
    }
  }
}

extension WorkOSAuthProvider: ASWebAuthenticationPresentationContextProviding {
  func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
    let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
    let windows = scenes.flatMap(\.windows)
    if let window = windows.first(where: \.isKeyWindow) ?? windows.first {
      return window
    }
    guard let scene = scenes.first else {
      preconditionFailure("Web authentication requires an active window scene")
    }
    return UIWindow(windowScene: scene)
  }
}

File: android-native/app/src/main/java/app/dimo/android/auth/WorkOSAuthProvider.kt
package app.dimo.android.auth

import android.content.Context
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import app.dimo.android.app.AppConfig
import dev.convex.android.AuthProvider
import kotlin.math.min
import kotlin.math.pow
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.Json

/**
 * Port of `ios-native/Dimo/Auth/WorkOSAuthProvider.swift`, implementing the
 * Convex Android `AuthProvider` contract.
 *
 * Session resilience: transient refresh failures are retried and never clear
 * the stored refresh token; only terminal `invalid_grant` signs the user out.
 */
class WorkOSAuthProvider(context: Context) : AuthProvider<WorkOSSession> {
  private val appContext = context.applicationContext
  private val store = TokenStore(appContext)
  private val json = Json { ignoreUnknownKeys = true }
  private val lock = Any()
  private val refreshMutex = Mutex()

  @Volatile
  private var cached: WorkOSSession? = null

  private var onIdToken: ((String?) -> Unit)? = null

  val currentAccessToken: String? get() = synchronized(lock) { cached?.accessToken }
  val currentSession: WorkOSSession? get() = synchronized(lock) { cached }
  val hasPersistedRefreshToken: Boolean get() = store.refreshToken != null

  /**
   * Cold-start restore. Retries transient failures; clears storage only on a
   * terminal WorkOS session end.
   */
  suspend fun restoreSession(): WorkOSSession? {
    if (!hasPersistedRefreshToken) return null
    repeat(5) { attempt ->
      try {
        return loginFromCache {}.getOrThrow()
      } catch (error: AuthException) {
        if (error.isTerminalRefresh) {
          clearPersisted()
          return null
        }
        delay((min(8.0, 2.0.pow(attempt.toDouble())) * 1000).toLong())
      } catch (_: Throwable) {
        delay((min(8.0, 2.0.pow(attempt.toDouble())) * 1000).toLong())
      }
    }
    // Keep the refresh token so a later launch can succeed.
    return null
  }

  suspend fun signIn(provider: String = "GoogleOAuth"): WorkOSSession =
    performSignIn(appContext, provider)

  suspend fun signOut() {
    clearPersisted()
    onIdToken = null
  }

  /**
   * Returns a valid session, refreshing when the access token is within a minute
   * of expiry. [force] refreshes unconditionally.
   */
  suspend fun refreshIfNeeded(force: Boolean = false): WorkOSSession {
    val current = synchronized(lock) { cached }
    if (!force && current != null && current.expiresAt - System.currentTimeMillis() > EXPIRY_SKEW_MS) {
      return current
    }
    return refreshSingleFlight()
  }

  // MARK: AuthProvider

  override suspend fun login(
    context: Context,
    onIdToken: (String?) -> Unit,
  ): Result<WorkOSSession> = runCatching {
    this.onIdToken = onIdToken
    val session = performSignIn(context, "GoogleOAuth")
    onIdToken(session.accessToken)
    session
  }

  override suspend fun loginFromCache(
    onIdToken: (String?) -> Unit,
  ): Result<WorkOSSession> = runCatching {
    this.onIdToken = onIdToken
    val current = synchronized(lock) { cached }
    if (current != null && current.expiresAt - System.currentTimeMillis() > EXPIRY_SKEW_MS) {
      onIdToken(current.accessToken)
      return@runCatching current
    }
    val session = refreshSingleFlight()
    onIdToken(session.accessToken)
    session
  }

  override suspend fun logout(context: Context): Result<Void?> = runCatching {
    clearPersisted()
    onIdToken = null
    null
  }

  override fun extractIdToken(authResult: WorkOSSession): String = authResult.accessToken

  // MARK: Private

  private suspend fun refreshSingleFlight(): WorkOSSession = refreshMutex.withLock {
    val current = synchronized(lock) { cached }
    if (current != null && current.expiresAt - System.currentTimeMillis() > EXPIRY_SKEW_MS) {
      return current
    }
    val refreshToken = synchronized(lock) { cached?.refreshToken } ?: store.refreshToken
      ?: throw AuthException.NotAuthenticated
    try {
      val session = WorkOSAPI.refresh(refreshToken, AppConfig.workOSClientID)
      persist(session)
      onIdToken?.invoke(session.accessToken)
      session
    } catch (error: AuthException) {
      if (error.isTerminalRefresh) clearPersisted()
      throw error
    }
  }

  private suspend fun performSignIn(context: Context, provider: String): WorkOSSession {
    val verifier = PKCE.makeVerifier()
    val challenge = PKCE.challenge(verifier)
    val state = PKCE.makeState()
    val url = WorkOSAPI.authorizationURL(
      clientId = AppConfig.workOSClientID,
      redirectURI = AppConfig.workOSRedirectURI,
      state = state,
      codeChallenge = challenge,
      provider = provider,
    )

    val redirect = AuthRedirectBus.expectRedirect()
    launchCustomTab(context, url)
    val callback: Uri = redirect.await()

    if (callback.getQueryParameter("state") != state) throw AuthException.StateMismatch
    val code = callback.getQueryParameter("code") ?: throw AuthException.MissingCode

    val session = WorkOSAPI.exchangeCode(
      code = code,
      codeVerifier = verifier,
      clientId = AppConfig.workOSClientID,
      redirectURI = AppConfig.workOSRedirectURI,
    )
    persist(session)
    return session
  }

  private fun launchCustomTab(context: Context, url: Uri) {
    val intent = CustomTabsIntent.Builder()
      .setShowTitle(false)
      .setUrlBarHidingEnabled(true)
      .build()
    intent.intent.addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK)
    intent.launchUrl(context, url)
  }

  private fun persist(session: WorkOSSession) {
    store.refreshToken = session.refreshToken
    store.userJson = json.encodeToString(WorkOSUser.serializer(), session.user)
    synchronized(lock) { cached = session }
  }

  private fun clearPersisted() {
    store.clear()
    synchronized(lock) { cached = null }
  }

  private companion object {
    const val EXPIRY_SKEW_MS = 60_000L
  }
}
